presto-main/src/main/java/com/facebook/presto/execution/QueueableExecution.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.concurrent.SetThreadName;
+
+import java.util.List;
+import java.util.concurrent.Executor;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class QueueableExecution
+{
+    private final QueryExecution queryExecution;
+    private final List<QueryQueue> nextQueues;
+    private final ListenableFuture<?> listenableFuture;
+    private final Executor executor;
+    private final SqlQueryManagerStats stats;
+
+    public QueueableExecution(QueryExecution queryExecution, List<QueryQueue> nextQueues, Executor executor, SqlQueryManagerStats stats)
+    {
+        this.queryExecution = checkNotNull(queryExecution, "queryExecution is null");
+        this.nextQueues = ImmutableList.copyOf(checkNotNull(nextQueues, "nextQueues is null"));
+        this.executor = checkNotNull(executor, "executor is null");
+        this.stats = checkNotNull(stats, "stats is null");
+        SettableFuture<?> settableFuture = SettableFuture.create();
+        SqlQueryManager.addCompletionCallback(queryExecution, () -> settableFuture.set(null));
```

Comment:
this is an escaped "this" reference since the callback can happen on the auto generated anonymous inner callback class.

while it might technically be ok, it's a bit of an anti-pattern. so you might want to pass in the ListenableFuture directly to the constructor (having the callback registered outside of the constructor). You can do this either at the point you create this class, or create a static factory function in this class that creates it first (in that case this constructor should be private then)